sdk/azidentity/environment_credential.go

```diff
@@ -9,30 +9,38 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
 )
 
-// NewEnvironmentCredential creates an instance of the ClientSecretCredential type and reads credential details from environment variables.
+// NewEnvironmentCredential creates an instance that implements the azcore.TokenCredential interface and reads credential details from environment variables.
 // If the expected environment variables are not found at this time, then a CredentialUnavailableError will be returned.
 // options: The options used to configure the management of the requests sent to Azure Active Directory.
-func NewEnvironmentCredential(options *TokenCredentialOptions) (*ClientSecretCredential, error) {
+func NewEnvironmentCredential(options *TokenCredentialOptions) (azcore.TokenCredential, error) {
 	tenantID := os.Getenv("AZURE_TENANT_ID")
 	if tenantID == "" {
 		err := &CredentialUnavailableError{CredentialType: "Environment Credential", Message: "Missing environment variable AZURE_TENANT_ID"}
 		logCredentialError(err.CredentialType, err)
 		return nil, err
 	}
-
 	clientID := os.Getenv("AZURE_CLIENT_ID")
 	if clientID == "" {
 		err := &CredentialUnavailableError{CredentialType: "Environment Credential", Message: "Missing environment variable AZURE_CLIENT_ID"}
 		logCredentialError(err.CredentialType, err)
 		return nil, err
 	}
-
-	clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
-	if clientSecret == "" {
-		err := &CredentialUnavailableError{CredentialType: "Environment Credential", Message: "Missing environment variable AZURE_CLIENT_SECRET"}
-		logCredentialError(err.CredentialType, err)
-		return nil, err
+	if clientSecret := os.Getenv("AZURE_CLIENT_SECRET"); clientSecret != "" {
+		azcore.Log().Write(LogCredential, "Azure Identity => NewEnvironmentCredential() invoking ClientSecretCredential")
+		return NewClientSecretCredential(tenantID, clientID, clientSecret, options)
 	}
-	azcore.Log().Write(LogCredential, "Azure Identity => NewEnvironmentCredential() invoking ClientSecretCredential")
-	return NewClientSecretCredential(tenantID, clientID, clientSecret, options)
+	if clientCertificate := os.Getenv("AZURE_CLIENT_CERTIFICATE_PATH"); clientCertificate != "" {
+		azcore.Log().Write(LogCredential, "Azure Identity => NewEnvironmentCredential() invoking ClientCertificateCredential")
+		return NewClientCertificateCredential(tenantID, clientID, clientCertificate, nil, options)
+	}
+	if username := os.Getenv("AZURE_USERNAME"); username != "" {
+		if password := os.Getenv("AZURE_PASSWORD"); password != "" {
+			azcore.Log().Write(LogCredential, "Azure Identity => NewEnvironmentCredential() invoking UsernamePasswordCredential")
+			return NewUsernamePasswordCredential(tenantID, clientID, username, password, options)
+		}
+	}
+	err := &CredentialUnavailableError{CredentialType: "Environment Credential", Message: "Missing environment variable AZURE_CLIENT_SECRET or AZURE_CLIENT_CERTIFICATE_PATH or AZURE_USERNAME and AZURE_PASSWORD"}
```

Comment:
We already log the env vars found before this right, so between that and this error message is should be clear what's missing?

Reply:
Yes but we also log when a credential returns an error and this is the error returned from NewEnvironmentCredential. I do think it's hard to miss it, particularly if the user is checking the error that is returned, but adding the error to the log completes the story for this credential dont you think?

Reply:
So we log what was found, and here (below really) we log what was missing so it sounds like the caller should be able to figure out from the logs which env vars they need to add.

Reply:
No keep logging the error.  I just want to ensure that the log messages are useful in tracking down what env vars are missing.  E.g. I'm using client secret env vars, if the previous log message states that it found the tenant, subscription, and client ID env vars, and this error logs that client secret (and other) env vars are missing it should be clear that I omitted (or misspelled) the env var for the client secret.

Reply:
Ah I understood you know 😅. Yes before we log the env vars that are found and here we log those that are missing. Do you think I should expand on the message? Saying please make sure you include one of these and that the env var is spelled correctly? Something along those lines

Reply:
Ok so it sounds like we both agree the logs produce enough info to help diagnose this error.

I'm fine with the message as-is but feel free to modify it if you think it would help.

Reply:
My only doubt with the error message as is, is if it's clear enough that AZURE_USERNAME and AZURE_PASSWORD both need to be present at the same time?

Reply:
Good point.  Maybe customize the message for that case.